File: Kotlin/Benchmark/src/main/kotlin/benchmark/mandelbrot.kt
/* mandelbrot.kt
 * Calculate Mandelbrot sets in the Kotlin
 * (http://kotlinlang.org) programming language
 */

package benchmark.mandelbrot

import complex.Complex

// first we need an container for the image data
data class Image<T>(val width: Int, val height: Int, val channels: Int,
  val data: Array<T>)

// calculates pixel values
fun pixelVal(z0: Complex, nIter: Int = 255, zMax: Double = 2.0): Int {
  var iter = 0
  var z = Complex.zero

  while(iter <= nIter) {
    if (z.abs() > zMax) return iter
    z = z *z + z0
    iter++
  }

  return nIter
}

// generates Mandelbrot set for given coordinates
fun mandelbrot(width: Int, height: Int,
  xCenter: Double, yCenter: Double,
  pixSize: Double = 4.0 / width): Image<Int> {

  val imgData = Array(width * height * 3, {_ -> 0})
  return Image(width, height, 3, imgData)
}

// prints an ASCII Mandelbrot set
/*for (i in -40.0..40.0) {
    for (r in -40.0..40.0) {
        print(mandelbrot(Complex(r - 25.0, i) / 35.0, 256)
                ?.let { 'a' + (it % 26) }
                ?: ' '
        )
    }
    println()
  }*/

File: Ruby/Benchmark/lib/Benchmark/mandelbrot.rb
# mandelbrot.rb
# Calculate Mandelbrot sets in Ruby

require 'complex'

module Benchmark

  class Mandelbrot
    # extend self
    attr_reader :width, :height

    # some constants
    R_MAX = 2.0
    N_MAX = 255

    # generate Mandelbrot set for the given coordinates
    def initialize(width, height, x_center, y_center, pixel_size)
      @width = width
      @height = height

      x_offset = x_center - 0.5 * pixel_size * width
      y_offset = y_center + 0.5 * pixel_size * height

      @image_data = Array.new(width * height) do |i|
        x = (i % width) * pixel_size + x_offset
        y = (i / width) * pixel_size - y_offset
        to_rgb(pixel_value(x, y))
      end
    end

    # calculate pixel value
    def pixel_value(x, y)
      n = 0
      z = Complex(0, 0)
      z0 = Complex(x, y)
      while z.abs < R_MAX && n < N_MAX
        z = z * z + z0
        n += 1
      end

      n
    end

    # convert pixel value to RGB
    def to_rgb(value)
      rgb = [5 * (value % 15), 32 * (value % 7), 8 * (value % 31)]

      rgb
    end

    # write image data to PPM file
    def to_ppm(file_name)
      File.open(file_name, 'w') do |file|
        file.write("P3\n")
        file.write("#{@width} #{@height} 3\n")
        @image_data.each do |pix|
          file.write("#{pix[0]} #{pix[1]} #{pix[2]}\n")
        end
      end
    end
  end
end
